src/bidiMapper/scripts/eval.es

```diff
@@ -16,117 +16,147 @@
+    } else if (value instanceof Object) {
+      // TODO: Recursive serialize.
+      let objectId = serializationMapper._objectToId.get(value);
+      if (!objectId) {
+        objectId = uuid();
+        serializationMapper._objectToId.set(value, objectId);
+        serializationMapper._idToObject.set(objectId, new WeakRef(value));
+      }
+
+      const result = { objectId };
+
+      if (typeof value === 'function') {
+        result.type = 'function';
+        return result;
+      }
+
+      result.type = 'object';
+      return result;
+    } else {
+      throw new Error('not yet implemented');
     }
+  }
 
-    deserialize(value) {
-      switch (value.type) {
-        case 'undefined': {
-          return undefined;
-        }
-        case 'null': {
-          return null;
-        }
-        case 'string': {
+  function deserialize(value) {
+    switch (value.type) {
+      case 'undefined': {
+        return undefined;
+      }
+      case 'null': {
+        return null;
+      }
+      case 'string': {
+        return value.value;
+      }
+      case 'number': {
+        if (value.value === 'NaN') {
+          return NaN;
+        } else if (value.value === '-0') {
+          return -0;
+        } else if (value.value === '+Infinity') {
+          return +Infinity;
+        } else if (value.value === '-Infinity') {
+          return -Infinity;
+        } else {
           return value.value;
         }
-        case 'number': {
-          if (value.value === 'NaN') {
-            return NaN;
-          } else if (value.value === '-0') {
-            return -0;
-          } else if (value.value === '+Infinity') {
-            return +Infinity;
-          } else if (value.value === '-Infinity') {
-            return -Infinity;
-          } else {
-            return value.value;
-          }
-        }
-        case 'boolean': {
-          return value.value;
+      }
+      case 'boolean': {
+        return value.value;
+      }
+      case 'function':
+      case 'object':
+        {
+        const weakRef = serializationMapper._idToObject.get(value.objectId);
+        if (!weakRef) {
+          throw new Error('unknown object reference');
         }
-        case 'object': {
-          const weakRef = this._idToObject.get(value.objectId);
-          if (!weakRef) {
-            throw new Error('unknown object reference');
-          }
-
-          const obj = weakRef.deref();
-          if (!obj) {
-            throw new Error('stable object reference');
-          }
-
-          return obj;
+
+        const obj = weakRef.deref();
+        if (!obj) {
+          throw new Error('stable object reference');
         }
-        default:
-          throw new Error('not yet implemented');
+
+        return obj;
       }
+      default:
+        throw new Error('not yet implemented');
+
     }
   }
 
   function uuid() {
-    return crypto.randomUUID();
+    // TODO sadym: `crypto.randomUUID()` works only in secure context.
```

Comment:
I tried configuring rollup and typescript to build this file so that the `uuid` node module could be imported. I ran into various problems trying to get this to work. Rollup outputs an IIFE with a semicolon at the end, and the semicolon causes a syntax error when wrapping the script text. This might be worth another look if we can't use crypto.randomUUID.

Reply:
I created a ticket for that: #49